Add annotation reference to fragment controller documentation

fritzmg opened the pull request:
Annotations were only mentioned as a side note so far. This adds a more complete reference to those articles.

fritzmg committed:
add annotation reference to fragment controller documentation

Toflar reviewed (dismissed):
Nice one!

aschempp commented on docs/dev/framework/content-elements.md:
```diff
@@ -143,6 +143,47 @@ $GLOBALS['TL_LANG']['CTE']['my_content_element'] = [
+
+The following example sets the type of the content element to `my_example`, puts 
+it in the `media` category, sets the template name to `ce_some_example` and defines
+the renderer to be `inline` (which is the default):
```
That's not exactly true, Contao ships with its own renderer (`forward`). I would either mention `esi` or none at all, as most users should not use that anyway 🙈

fritzmg replied:
Hmmm... I am confused now, in the core `inline` is used as well as `forward`.

aschempp replied:
Not sure what you're referring to? The default renderer for Contao fragments is `forward`: https://github.com/contao/contao/blob/master/core-bundle/src/DependencyInjection/Compiler/RegisterFragmentsPass.php#L106

fritzmg replied:
In the FragmentHandler the default is `inline`:

https://github.com/contao/contao/blob/fd61bc4e34608f38112f97dfa1723337362ba229/core-bundle/src/Fragment/FragmentHandler.php#L66

But yeah, when registering a fragment controller, it would default to `forward` obviously. However, that means that the possible values are `forward`, `inline` and `esi`, correct?

aschempp replied:
there's more than these three (like `hinclude` or any custom one). But these three are the only ones that make sense. Or maybe `inline` never makes sense now that we have `forward` 🤷‍♂

aschempp commented on docs/dev/framework/content-elements.md:
```diff
@@ -143,6 +143,47 @@ $GLOBALS['TL_LANG']['CTE']['my_content_element'] = [
+
+Instead of tagging the content element controller service via the service configuration,
+the service tag can also be configured through annotations, as already used in the 
+code example above. The annotation can be used on the class of the content element 
```
> can be used on the class

if that class is invokable (has `__invoke` method), or basically if it extends the abstract controller

fritzmg committed:
annotations on invokable classes

fritzmg committed:
expand on fragment renderers, caching and ESI

fritzmg commented:
To make the `renderer` options more clear, I have expanded this PR with the following:

* All three renderer options mentioned.
* Added a new sections in the Caching documentation about fragments & ESI.

I also added a brief note mentioning how to implement content elements and front end modules in previous Contao versions.

@Toflar @aschempp pls review :)

Toflar commented on docs/dev/framework/caching.md:
```diff
@@ -215,8 +217,27 @@ Moreover, you don't have to register to all the different callbacks such as `ons
+can provide its own cache tags, as mentioned previously.
+
+Contao brings its own `forward` fragment renderer, which provides the fragment with 
+a full clone of the request (contrary to Symfony's default `inline` renderer). The 
```
The info missing here is that it will also provide the `POST` data, in contrary to the `inline` renderer.

fritzmg replied:
I have updated the wording.

Toflar reviewed (commented):
At some point, I would also like to make some matrix that should help the developers understand when to choose ESI :) It's a really difficult topic to understand. The fact that it's a fragment is easy to grasp but the implications aren't.

fritzmg commented:
> At some point, I would also like to make some matrix that should help the developers understand when to choose ESI :) It's a really difficult topic to understand. The fact that it's a fragment is easy to grasp but the implications aren't.

Agreed. The documentation is a living thing anyway :)

fritzmg committed:
provide more information about the forward renderer
